### ISSUE TITLE
Working with virtual environments (Part 2)

### ISSUE THREAD

**Issue by aes512** [rundeck-plugins/ansible-plugin#51]
Being able to set an arbitrary virtualenv path would be huge. I have seen several use cases where there are multiple development teams that are authoring Ansible playbooks, over time there is a big gap between the versions that each of the teams are using.

The development teams will often not have the bandwidth to refactor their playbooks to support a specific (backwards-INcompitable) version of Ansible. This leads to an organization being stuck on an ancient version of Ansible configured on the Rundeck server - therefore making the usage of this plugin impossible (jobs then have to be managed with wrapper scripts to set the environments).

**Comment by frozenice** [comment]
I don't use Python and virtualenv, what would the plugin need to do before launching Ansible?

**Comment by yassineazzouz** [comment]
Hi @aes512 ,

My understanding is that you use multiple python virtualenv with different ansible versions in it ? and you would like to be able to specify the  virtualenv name as part of the project configuration ? I think you can still do this manually by switching the environment in the server, right ?

**Comment by aes512** [comment]
@yassineazzouz yes. Multiple projects and jobs, each having separate Python virtual environments. All we have done thus far to compliment this environment is to wrap the ansible/ansible-playbook commands in a shell script, which (1) sources the virtualenv and (2) runs the ansible/ansible-playbook command(s).

The Rundeck server itself cannot be tied to a specific version of Ansible due to the fact that there are several different Ansible versions that need to be supported. I am not sure what you mean about _switching the environment in the server_?

**Comment by yassineazzouz** [comment]
Hi @aes512 , I see so the environment get passed to ansible through the shell parent process. Well what I meant is just sourcing the virtualenv in /etc/rundeck/profile, but that will require you to restart rundeckd each time you source a new file, which I guess is a bit painful. 
I can't think of a way of doing this, other than sourcing the virtualenv in the plugin itself. right ?

**Comment by DazWorrall** [comment]
The ability to specify the path to the ansible executable should be enough, you could pick the script installed in the virtualenv.

**Comment by kbenyous** [comment]
Hi all,

I need the same feature : 
- select the Python version,
- select the Ansible version,
- set the Python modules, Ansible plugins and roles, etc. before  running my playbook. They are playbook-specific, then job-specific.  I do this with pip (for python modules), and ansible-galaxy (for Ansible roles). 

Each Ansible playbook and each rundeck job has his specific requirements, and I would like Ansible to run in his own isolated virtualenv. 

A nice workflow would be : 
1. Run a workflow step that creates a fresh virtualenv. The step basically needs the Python version and the Ansible version. It could be implemented with a simple shell script (job argument?). Then export the location of the activation script for the virtualenv.
2. Run a workflow step that setups the virtualenv with pip and ansible-galaxy, according to a requirements list.  It could be implemented with a simple shell script (job argument?).
3. Run the virtualenv activation script and ansible-playbook
4. Delete the virtualenv

The point is that I don't know how to export the location of the virtualenv activation script from 1. to 3. :-(

**Comment by aes512** [comment]
@DazWorrall  where would I do that exactly? 

> > DazWorrall commented on Aug 8
> > The ability to specify the path to the ansible executable should be enough, you could pick the script installed in the virtualenv.

**Comment by yassineazzouz** [comment]
Hi @kbenyous ,

**Comment by bbhenry** [comment]
+1

**Comment by mcassaniti** [comment]
Your PR is most welcome for this request, but personally it isn't something I'd be implementing anytime soon. It's also potentially confusing. There's already a lot of options exposed to the end user, but providing the executable path to ansbile and ansible-playbook per job is getting a bit extreme.

**Comment by joshbenner** [comment]
Setting a virtualenv rather than simply path to executable is advantageous because any external plugins that ansible invokes (inventory, etc.) will not necessarily use the the same python environment as the ansible binary triggered.

This is because ansible doesn't assume plugins are python, and so essentially makes a shell call to run them. With virtualenv, the called plugins will naturally use that same environment.

**Comment by mcassaniti** [comment]
@joshbenner I certainly understand the benefit of using virtualenv and there's no objection to it. You're most welcome to submit a PR for it (hope that doesn't sound rude).

**Comment by dangoncalves** [comment]
What if we just ask the virtualenv PATH to the user and we set the environment variables VIRTUALENV and PATH according to whate the activate script do? Then we can transmit the environment to the ansible command and voilà.

**Comment by dangoncalves** [comment]
What if we set the VIRTUALENV and PATH environment variables according to what the activate script do? Then we can pass the environment to the ansible command and voilà.

**Comment by fdevans** [comment]
Does PR #315 fix this?

**Comment by ditshe** [comment]
> Does PR #315 fix this?

sad to say, but no
my rundeck job log 
`Failed: IOFailure: ERROR: Ansible IO failure: Cannot run program "/bin/ansible-playbook" (in directory "/home/nny/venv/ansible"): error=2, No such file or directory`

my pwd/ls
```shell
$ pwd
/home/nny/venv/ansible
~/venv/ansible/bin$ ls | grep play
ansible-playbook
``` 
but I can manually run
`ANSIBLE_ROLES_PATH=/home/nny/venv/ansible/roles /home/nny/venv/ansible/bin/ansible-playbook -i /home/nny/venv/ansible/hosts /home/nny/venv/ansible/playbooks/config/build.pb.yml`

**Comment by fdevans** [comment]
Sorry we missed this. If it's still a problem please open a new Issue with current Rundeck version details, and a deeper description of how this can be repeated.
